File: priceserver/modules/sdr/currencylayer.py
import requests

from typing import Dict
from modules import settings

CL_ACCESS_KEY = settings['CURRENCYLAYER_API_KEY']
CL_ENDPOINT = "https://apilayer.net/api/live"

latest = None


def get_sdr_rates(currencies) -> Dict[str, float]:
    global latest

    rates = {}
    url = f"{CL_ENDPOINT}?access_key={CL_ACCESS_KEY}&source=XDR&currencies={','.join(currencies)}"
    try:
        res = requests.get(url).json()
        if not res["success"]:
            raise RuntimeError

        for quote, rate in res["quotes"].items():
            code = quote[3:]
            rates[code] = 1/float(rate)

    except requests.exceptions.BaseHTTPError:
        if not latest:
            raise RuntimeError

        return latest

    latest = rates
    return rates

File: priceserver-ts/src/provider/fiat/CurrencyLayer.ts
import nodeFetch from 'node-fetch';
import { errorHandler } from 'lib/error';
import { toQueryString } from 'lib/fetch';
import { Quoter } from '../base';

interface Response {
  success: boolean;
  quotes: { [symbol: string]: number };
  error?: any;
}

export class CurrencyLayer extends Quoter {
  private async updatePrices(): Promise<void> {
    const params = {
      access_key: this.options.apiKey,
      source: this.baseCurrency,
      currencies: this.quotes.map(quote => (quote === 'SDR' ? 'XDR' : quote)).join(',')
    };

    const response: Response = await nodeFetch(`https://apilayer.net/api/live?${toQueryString(params)}`, {
      timeout: this.options.timeout
    }).then(res => res.json());

    if (!response || !response.success || !response.quotes) {
      throw new Error(`wrong response, ${response}`);
    }

    // update last trades
    for (const symbol of Object.keys(response.quotes)) {
      const quote = symbol.replace('KRW', '');
      this.priceByQuote[quote === 'XDR' ? 'SDR' : quote] = +response.quotes[symbol];
    }
  }

  protected async update(): Promise<boolean> {
    await this.updatePrices().catch(errorHandler);

    return true;
  }
}

export default CurrencyLayer;
